docs/enhancements/fep-001-api-fleet-templating.md

```diff
@@ -0,0 +1,300 @@
+as devices will render the TemplateVersions in different ways over time, we probably
+need an audit log where we can record how devices are being rendered over time.
+
+Would it suffice to have the ability to send those specific traces to a logging system?
+Keeping those in the database would probably be dangerous over time, specially at the scales
+we intend this system to work.
+
+
+### Risks and Mitigations
+
+When rolling out a configuration to a fleet of devices, during the rollout, some devices could turn out to be
+not renderable, for example:
+* due to a missing label
+* because the template rendering is referencing non existing resources
+
+This is partially mitigated by the initial device scan performed by the fleet controller
+when a fleet.spec.template changes. Still, labels could change while a fleet is being rolled out.
+
+We probably need to differentiate between "fleet controller did not find the label, so
+failed to do the substitution" and "the referenced resource (after substitution) does
+not exist / isn't accessible". There is little the service can do about the latter:
+verifying service-side is costly and also futile, because the resource could be there when
+the service tried, but no longer there when the agent needs it. At the end of the day,
+the agent will try during staging, fail, and set its condition to "failed to reconcile".
+The former case would be similar (failure to reconcile). Then, if errors exceed threshold,
+stop the rollout.
+
+How do we signal this is happening to a device? Status on the device is only set by the
+agent.
+
+## Design Details
+
+TBD
+
+### Scalability
+
+TBD
+
+### Troubleshooting
+
+Troubleshooting can be done by looking at the device status and rendered device spec. Any errors found in templates
+could be reported as part of the fleet status conditions.
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
+
+### Every configuration combination has it's own fleet
+This is the most obvious alternative, this proposal is not implemented, and administrators
+take care of replicating fleet definitions as necessary to accomodate the different
+configurations.
+
+This approach does not allow customization based on fields like the device name.
+
+### Fleet template inheritance / sub-fleets
+Fleets can inherit base configuration from other fleets and modify only the necessary changes
+we then use labels to attach devices to specific sub-fleets.
+
+This approach is less flexible and does not allow customization based on fields like the device name.
+
+### Template conditionals
+
+Instead of using variables, allow optional config items in the fleet template. 
+Something like "if label key equals value, then include this secret".  Then we
+can "freeze" the entire template because we know all of the options ahead of time.
+
+```yaml
+kind: Fleet
+metadata:
+  name: autonomous-forklifts
+spec:
+  template:
+    spec:
+      os:
+        switchFromLabel:
```

Comment:
wouldn't a mutation of the spec be enough to show intent? It's not clear to me what ` switchFromLabel` is attempting.

Reply:
I see now so conditional image spec.

Reply:
Would it make sense to use a labelSelector then template the device spec?

Reply:
I think the idea was to make the system able to pre-calculate, I think this alternative does not need a lot of development now, as @avishayt had some agreement on the feasibility of the main one now (based on  #247 final proposal).

I am happy to continue developing the alternative if we don't see the main one going forward.